DSEGOG-8 user can reorder columns

sam-glendenning opened the pull request:
## Description
Note: this contains resizing columns functionality so probably best to review that PR first - https://github.com/ral-facilities/operationsgateway/pull/4

Adding functionality to allow the user to reorder the table columns using drag and drop functionality provided by [react-beautiful-dnd](https://github.com/atlassian/react-beautiful-dnd). The result is a nice looking, simplistic method by which the user can reorder the displayed columns without having to add or remove columns to achieve the same thing. To drag and drop, the user can click anywhere on the column header that isn't the close button or the divider. I had to be a bit clever to make sure their functionality didn't overlap.

It's not really feasible to test this functionality outside of a browser so I used this opportunity to add Cypress tests to the project. I've only done tests for this functionality so far, not made any for the previous functionality I've done yet. Currently using Cypress 9.6.0 to match DataGateway because this was the simplest to get set up with. I considered upgrading to Cypress v10 but this contains many breaking changes so I made a separate issue (stfc.atlassian.net/jira/software/projects/DSEGOG/issues/DSEGOG-33) for this to tackle later.

## Testing instructions
Experiment with adding columns to the display and then reordering them by dragging the headers. Experiment with both large and small tables to make sure the functionality works in all cases. Also make sure that the drag and drop functionality does not interfere with column resizing and closing.

- [x] Review code
- [x] Check Actions build
- [x] Review changes to test coverage

## Agile board tracking
Closes [DSEGOG-8](https://stfc.atlassian.net/jira/software/projects/DSEGOG/boards/22?selectedIssue=DSEGOG-8)

Commit:
DSEGOG-8 adding basic drag and drop reordering

This is done through react-beautiful-dnd. Currently, columns can only be moved to the beginning or end of the column list. The order also gets messed up when a column is added or removed

Commit:
DSEGOG-8 updating column order using table state

Making use of useColumnOrder allows for more native management of column ordering

Commit:
DSEGOG-8 prevent reordering when user is resizing

This fixes a bug which caused a column to be dragged whenever it is resized. It was no use trying to disable dragging whenever the column.isResizing field is flagged as true because this only happened after the mouse event triggered, so was always too late. So we use onMouseOver and onMouseOut events instead to enable and disable the behaviour in time.

I spent too long trying to figure out a solution to this

Commit:
DSEGOG-8 fixing bug with column order not being correctly set

This was caused by accidentally making use of displayedColumns instead of visibleColumns. This effectively meant that the column order could only be changed once before failing.

Commit:
DSEGOG-8 adding unit and browser-based tests

This ensures drag and drop functionality can be properly tested through Cypress

Commit:
DSEGOG-8 Updating .gitignore for Cypress files and introduced e2e tests into CI build

Commit:
DSEGOG-8 removing react-dnd dependencies

These were used during experimentation and are no longer needed

sam-glendenning commented on src/table/headerRenderers/dataHeader.component.tsx:
```diff
@@ -84,58 +92,82 @@ const DataHeader = (props: DataHeaderProps): React.ReactElement => {
+                display: 'flex',
+                flexDirection: 'row',
+              }}
+              onMouseOver={() => {
+                setPermitResizing(false);
+              }}
+              onMouseOut={() => {
+                setPermitResizing(true);
+              }}
```
Most of the loss of codecov here is over these two events. Any advice on how to test these (or even if we should) is appreciated.

sam-glendenning commented on src/table/table.component.tsx:
```diff
@@ -65,10 +76,54 @@ const Table = React.memo((props: TableProps): React.ReactElement => {
+
+  const updater = (result: any): string[] => {
+    // Unlike displayedColumns, visibleColumns retains the current column order
+    const items = Array.from(visibleColumns);
+    const [reorderedItem] = items.splice(result.source.index, 1);
+    items.splice(result.destination.index, 0, reorderedItem);
+
+    return items.map((column: Column) => {
+      return column.Header?.toString() ?? '';
+    });
+  };
 
-  const { getTableProps, getTableBodyProps, headerGroups, rows, prepareRow } =
-    tableInstance;
+  const handleOnDragEnd = (result: DropResult): void => {
+    if (!result.destination) return;
+    setColumnOrder(updater(result));
+  };
```
Again, loss of codecov here. Not even sure how to trigger this behaviour in a unit test. Worst comes to worst we can override the codecov threshold and just merge in?

Commit:
DSEGOG-8 renaming state variable

Just realised I got confused and named the dataHeader state variable the wrong thing. We're not trying to permit resizing - we're trying to permit dragging

sam-glendenning commented:
Noticed a small bug with column ordering when columns are added/removed

Commit:
DSEGOG-8 moving checkboxes to table component, fixing column ordering issues

Moving the column checkboxes to be part of the Table component. This makes sense, since we always want the checkboxes to appear alongside the Table and their content should be defined by the Table.

Also redefining column structures and action handlers to better represent user actions and intent. Overall, this improves maintainability and fixes the bug we had whereby columns were incorrectly ordered when a column was removed and added back in.

Commit:
DSEGOG-8 updating tests

Commit:
Merge branch 'develop' into DSEGOG-8-user-can-reorder-columns

Commit:
DSEGOG-8 updating snapshots after develop merge

sam-glendenning commented:
@louise-davies apologies for the length this PR has turned out as but it is now ready for review again. It includes a lot of optimisations that improve the maintainability of things going forward. 

The gist of what I've done since I marked this as draft is fixing a bug which meant, when removing a column and adding it back, it went back into the same place it was removed from instead of being added to the end of the table as expected. I also added some code that more efficiently manages the column order and the table state overall. Please try and test as much of the functionality as possible to make sure everything plays nicely with each other.

Although this is a lengthy PR, remember that much of it is also taken up by yarn package additions and a couple of big snapshots!

louise-davies commented on src/table/headerRenderers/dataHeader.component.tsx:
```diff
@@ -87,64 +92,85 @@ const DataHeader = (props: DataHeaderProps): React.ReactElement => {
+              <div
+                {...resizerProps}
+                style={{
+                  marginLeft: 10,
+                  paddingLeft: '4px',
```
This I think was a merge conflict resolution gone wrong - as adding this back in has reverted the fix I made in the resizing PR. I shall fix this as I've already done the code - just need to update the snapshots in a sec

sam-glendenning replied:
Thanks Louise, looks like I got confused

louise-davies commented on src/table/table.component.tsx:
```diff
@@ -28,24 +36,37 @@ export interface TableProps {
+   ** availableColumns - these are the columns passed in from RecordTable.
+   **                    They represent all columns that can currently be added to the
+   **                    display based on data received from the backend
+   ** checkedColumns   - these are the columns the user has selected to appear in the table
```
Minor point, but I dislike the `checkedColumns` name as I think it's too based on the WIP column checkboxes selector, rather than what it will look like in the future. Maybe `selectedColumns` makes more sense? I'm not sure I care enough for you to change it though 😅

sam-glendenning replied:
I also prefer selectedColumns. I can change in a later PR since I'm working on a new feature right now

louise-davies commented on src/table/table.component.tsx:
```diff
@@ -78,13 +128,43 @@ const Table = React.memo((props: TableProps): React.ReactElement => {
+
+  const { columnOrder } = state;
+
+  const updater = (result: any): string[] => {
```
The name of this function is very generic to the point it's hard to tell what it's for. Can we rename this something more verbose? colOrderUpdater or something??

louise-davies committed:
DSEGOG-8 fix table borders via adjusting padding

louise-davies commented:
@sam-glendenning with the commit I've just done I just removed the right padding on table cells which fixes that overflow. I don't think we need the right padding but we can see later with real data if the text going right up to the border isn't wanted

louise-davies commented on src/table/table.component.tsx:
```diff
@@ -168,6 +262,12 @@ const Table = React.memo((props: TableProps): React.ReactElement => {
             page={page}
             rowsPerPage={resultsPerPage}
           />
+          <ColumnCheckboxes
+            availableColumns={availableColumns}
+            checkedColumns={checkedColumns}
+            onColumnOpen={handleColumnOpen}
+            onColumnClose={handleColumnClose}
+          />
```
Should this really be in the table component? In the future, this will be a separate epic and thus I don't think it should be in `Table`. However, by then I'd expect we'd have redux and perhaps that will result in this not having to be nested in `Table` as well...

sam-glendenning replied:
My reasoning is that we can't have one without the other so they're intrinsically linked. If we have checkboxes, even if none of them are selected, we still have a table element and the table will always have an ID column loaded, for example

louise-davies replied:
true, I suppose we'll see how the structure pans out when we redux everything & we refactor `ColumnCheckboxes` into the channel selector

Commit:
DSEGOG-8 addressing PR comments

Renaming checkedColumns to selectedColumns

Renaming updater function to more descriptive columnOrderUpdater

louise-davies reviewed (approved):
Functionality works well and looks nice plus I can still access column closing & resizing.
